### ISSUE TITLE
Actors not persisting state

### ISSUE THREAD

**Issue by antogh** [Azure/service-fabric-issues#396]
I have an application using the service fabric actors. 

I love the SF actors programming model, it suits very well my needs for this application, but unfortunately I have a problem with actors state persistence which even after many attempts I am not able to solve, hence I kindly ask your help to understand what is going wrong. 

The application is a sort of social network where each user is mapped into a SF actor and of course user preferences (like personal interests, interactions with the community, blocked users and so on) are stored into the actor state. I keep just minimal 'hot' data into state, cold data is going to cheaper storage services (db or blob).

The application logic works well, I ran it on a single dev machine with the local 5 node dev cluster and I could manage a dozen of users interacting between themselves independently in a very smooth and fast way. 

The problem occured after a couple of hours of inactivity from the tests, the users lost their preferences, in technical terms, the actors representing the users lost their state. I can say that because the application has a feature that shows user preferences and when you login for the first time it gives a short welcome intro and invites you to fill in your preferences, after you do that, when you come back, the intro is skipped and current preferences are listed. 

I tried to reproduce the problem and saw that after a couple of hours (more or less) of inactivity the actors always lose their state. The computer does not go in sleep mode during this time. 

Then I found a way to reproduce quickly the problem: In SF explorer I restart the node where the user actor primary replica is located. That node become quickly a secondary replica and the primary replica is moved onto another node. That's fine as a node failover process but the side effect is that the state of my actors is lost! :( It's if like the state data was not moved from the failing primary replica to the new primary replica, but that's strange, the purpose of replica should be to recover the state on failures.

Now some more tech details. 

- The actor keeping the users state is declared with the following attribute
   [StatePersistence(StatePersistence.Persisted)]
  (even though the application beahaves as if it was Volatile).

-  I debugged the application to see what happens when it lose the state (triggered restarting the primary replica node). As soon as an actor method is called  the OnActivateAsync is called and reset the state to its default. Each time an actor  method that modify the state is invoked it loads the state on entry and saves it before exiting. This is a snippet from the load/save function, it seems to work always without errors or exceptions:

```
      private async Task<bool> LoadCurState() {
         ConditionalValue<TgramUserState> cond_state = await StateManager.TryGetStateAsync<TgramUserState>(TgUConst.state_key);
         if (cond_state.HasValue)
            State = cond_state.Value;
         return cond_state.HasValue;
      }


      private async Task<bool> SaveCurState() {
         bool saved = false;
         try {
            await StateManager.SetStateAsync<TgramUserState>(TgUConst.state_key, State);
            saved = true;
         }
         catch (Exception e) {
            string err = e.Message;
         }
         return saved;
      }

```

- I thought the problem could be I was using VS2015 and sf actors SDK 2.0.217, i migrated the app to VS2017 and actors SDK 2.6.220. The problem remains the same :(

- I thought the problem could be I was using a local 5 nodes dev cluster, so I tried a real 5 node cluster on azure . The problem remains the same :( 

- I tried also the local 1 node dev cluster, it is still losing the state after a couple of hours :( 

I spent quite some time trying to solve this problem, needless to say the application is useless if it loses the state and I refuse to replace the state with a db because that would defeat the advantages of microservices. 

I ran out of ideas. Do you have any suggestion? 

In case you need it I could set up a VM in azure with the application always online so you can check the behavior. It can be used by any browser or in native android/ios app.

Thanks 
Antonio

**Comment by harahma** [comment]
@antogh Is it possible for you to share code in your actor class?

**Comment by antogh** [comment]
@harahma Thank you so much to you and the team for the quick reply and support. Yes I will share the code, I am trying to simplify the code and extract  only the part with the problem. I should have done this from the start, sorry. I Will update this thread asap. Best regards. Antonio

**Comment by antogh** [comment]
I think I found the problem in my code, I was not restoring properly the state in the OnActivateAsync, so when the primary node was moved (moved for failover or whatever reason) I was losing the state. 
I am doing the last checks and then I will close myself this issue in a couple of days, unless I discover some new problem, but I really don't think so.
I am really sorry I bothered you. 
I do appreciate a lot the prompt support.
Thanks 
Antonio
